ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py

```diff
@@ -0,0 +1,185 @@
+            InterpretedAudioUserIntent, self._expect_uintent_topic, 1
+        )
+        self._interp_publisher = self.create_publisher(
+            InterpretedAudioUserIntent, self._interp_uintent_topic, 1
+        )
+
+        self.message_queue = queue.Queue()
+        self.handler_thread = threading.Thread(target=self.process_message_queue)
+        self.handler_thread.start()
+
+    def listener_callback(self, msg):
+        """
+        This is the main ROS node listener callback loop that will process all messages received
+        via subscribed topics.
+        """
+        self.log.debug(f'Received message:\n\n"{msg.value}"')
+        self.message_queue.put(msg)
+
+    def process_message_queue(self):
+        """
+        Constant loop to process received messages.
+        """
+        while True:
+            msg = self.message_queue.get()
+            self.log.debug(f'Processing message:\n\n"{msg.value}"')
+            intent, score = self.detect_intents(msg)
+            if not intent:
+                continue
+            self.publish_msg(msg.value, intent, score)
+
+    def detect_intents(self, msg):
+        """
+        Keyphrase search for intent detection. This implementation does simple
+        string matching to assign a detected label.
+        """
+
+        def _tiebreak_intents(intents, confidences):
+            classification = intents[0]
+            score = confidences[0]
+            if len(intents) > 1:
+                for i, intent in enumerate(intents):
+                    if intent == INTENT_LABELS[2]:
+                        classification, score = intent, confidences[i]
+                self.log.info(
+                    f'Detected multiple intents: {intents}. Selected "{classification}".'
+                )
+            return classification, score
+
+        lower_utterance = msg.value.lower()
+        intents = []
+        confidences = []
+        if self._contains_phrase(lower_utterance, NEXT_STEP_KEYPHRASES):
+            intents.append(INTENT_LABELS[0])
+            confidences.append(0.5)
+        if self._contains_phrase(lower_utterance, PREV_STEP_KEYPHRASES):
+            intents.append(INTENT_LABELS[1])
+            confidences.append(0.5)
+        if self._contains_phrase(lower_utterance, QUESTION_KEYPHRASES):
+            intents.append(INTENT_LABELS[2])
+            confidences.append(0.5)
+        if not intents:
+            colored_utterance = colored(msg.value, "light_blue")
+            self.log.info(f'No intents detected for:\n>>> "{colored_utterance}":')
+            return None, -1.0
+
+        classification, confidence = _tiebreak_intents(intents, confidences)
+        classification = colored(classification, "light_green")
+        return classification, confidence
+
+    def publish_msg(self, utterance, intent, score):
+        """
+        Handles message publishing for an utterance with a detected intent.
+        """
+        intent_msg = InterpretedAudioUserIntent()
```

Comment:
Also fill in ROS header for this message.

Reply:
Not sure if I filled it out properly, but I think this  should be done.